File: js/src/core/operators.ts
/**
 * JSON-logic operators for the Tenet VM.
 * All operators are nil-safe: operations on nil/undefined return appropriate defaults.
 */

import type { EvalState } from './types.js';

/** Operator function signature */
type OperatorFn = (args: unknown, state: EvalState, resolve: ResolveFn) => unknown;
type ResolveFn = (node: unknown, state: EvalState) => unknown;

/**
 * Convert a value to a number if possible.
 */
export function toFloat(v: unknown): [number, boolean] {
    if (v === null || v === undefined) {
        return [0, false];
    }
    if (typeof v === 'number') {
        return [v, true];
    }
    if (typeof v === 'string') {
        // Don't auto-convert strings to numbers
        return [0, false];
    }
    return [0, false];
}

/**
 * Parse a date value (string or Date).
 * Supports ISO 8601 formats.
 */
export function parseDate(v: unknown): [Date, boolean] {
    if (v === null || v === undefined) {
        return [new Date(0), false];
    }
    if (v instanceof Date) {
        return [v, true];
    }
    if (typeof v === 'string') {
        const d = new Date(v);
        if (!isNaN(d.getTime())) {
            return [d, true];
        }
    }
    return [new Date(0), false];
}

/**
 * Determine if a value is "truthy" in JSON-logic terms.
 * nil, false, 0, and "" are falsy. Everything else is truthy.
 */
export function isTruthy(value: unknown): boolean {
    if (value === null || value === undefined) {
        return false;
    }
    if (typeof value === 'boolean') {
        return value;
    }
    if (typeof value === 'number') {
        return value !== 0;
    }
    if (typeof value === 'string') {
        return value !== '';
    }
    if (Array.isArray(value)) {
        return value.length > 0;
    }
    if (typeof value === 'object') {
        return Object.keys(value).length > 0;
    }
    return true;
}

/**
 * Compare two values for equality with type coercion.
 * nil == nil is true, nil == anything_else is false.
 */
export function compareEqual(a: unknown, b: unknown): boolean {
    if (a === null || a === undefined) {
        return b === null || b === undefined;
    }
    if (b === null || b === undefined) {
        return false;
    }

    // Try numeric comparison if both can be numbers
    const [aNum, aOk] = toFloat(a);
    const [bNum, bOk] = toFloat(b);
    if (aOk && bOk) {
        return aNum === bNum;
    }

    // String comparison
    return String(a) === String(b);
}

/**
 * Resolve arguments array, handling single values and missing args.
 */
function resolveArgs(args: unknown, expected: number, resolve: ResolveFn, state: EvalState): unknown[] {
    const result: unknown[] = new Array(expected).fill(null);

    if (!Array.isArray(args)) {
        // Single value case (e.g., {"not": true})
        if (expected > 0) {
            result[0] = resolve(args, state);
        }
        return result;
    }

    for (let i = 0; i < expected && i < args.length; i++) {
        result[i] = resolve(args[i], state);
    }
    return result;
}

/**
 * Get variable value from schema definitions or current element context.
 */
function getVar(path: string, state: EvalState): unknown {
    // Empty path returns current element context (for some/all/none)
    if (path === '') {
        return state.currentElement;
    }

    const parts = path.split('.');

    // Check definitions first
    const def = state.schema.definitions[parts[0]];
    if (def) {
        if (parts.length === 1) {
            return def.value;
        }
        // Nested access into the value
        return accessPath(def.value, parts.slice(1));
    }

    // Check derived state
    if (state.schema.state_model?.derived) {
        const derived = state.schema.state_model.derived[parts[0]];
        if (derived) {
            // Note: derived values should already be computed by this point
            // This is a fallback for direct access
            return undefined;
        }
    }

    return undefined;
}

/**
 * Access nested path in an object.
 */
function accessPath(value: unknown, parts: string[]): unknown {
    if (parts.length === 0 || value === null || value === undefined) {
        return value;
    }

    if (typeof value === 'object' && !Array.isArray(value)) {
        const obj = value as Record<string, unknown>;
        const next = obj[parts[0]];
        return accessPath(next, parts.slice(1));
    }

    return undefined;
}

// ============================================================
// Operator implementations
// ============================================================

const operators: Record<string, OperatorFn> = {
    // === Variable Access ===
    'var': (args, state) => {
        const path = typeof args === 'string' ? args : '';
        return getVar(path, state);
    },

    // === Comparison Operators ===
    '==': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        return compareEqual(a[0], a[1]);
    },

    '!=': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        return !compareEqual(a[0], a[1]);
    },

    '>': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return false;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        return aOk && bOk && aNum > bNum;
    },

    '<': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return false;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        return aOk && bOk && aNum < bNum;
    },

    '>=': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return false;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        return aOk && bOk && aNum >= bNum;
    },

    '<=': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return false;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        return aOk && bOk && aNum <= bNum;
    },

    // === Logical Operators ===
    'and': (args, state, resolve) => {
        if (!Array.isArray(args)) {
            return isTruthy(resolve(args, state));
        }
        for (const arg of args) {
            if (!isTruthy(resolve(arg, state))) {
                return false;
            }
        }
        return true;
    },

    'or': (args, state, resolve) => {
        if (!Array.isArray(args)) {
            return isTruthy(resolve(args, state));
        }
        for (const arg of args) {
            if (isTruthy(resolve(arg, state))) {
                return true;
            }
        }
        return false;
    },

    'not': (args, state, resolve) => {
        const a = resolveArgs(args, 1, resolve, state);
        return !isTruthy(a[0]);
    },

    '!': (args, state, resolve) => {
        const a = resolveArgs(args, 1, resolve, state);
        return !isTruthy(a[0]);
    },

    'if': (args, state, resolve) => {
        if (!Array.isArray(args) || args.length < 2) {
            return null;
        }

        // Process condition-then pairs
        for (let i = 0; i + 1 < args.length; i += 2) {
            const condition = resolve(args[i], state);
            if (isTruthy(condition)) {
                return resolve(args[i + 1], state);
            }
        }

        // Else clause (odd number of elements = has else)
        if (args.length % 2 === 1) {
            return resolve(args[args.length - 1], state);
        }

        return null;
    },

    // === Arithmetic Operators ===
    '+': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return null;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        if (!aOk || !bOk) {
            return null;
        }
        return aNum + bNum;
    },

    '-': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return null;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        if (!aOk || !bOk) {
            return null;
        }
        return aNum - bNum;
    },

    '*': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return null;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        if (!aOk || !bOk) {
            return null;
        }
        return aNum * bNum;
    },

    '/': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        if (a[0] === null || a[0] === undefined || a[1] === null || a[1] === undefined) {
            return null;
        }
        const [aNum, aOk] = toFloat(a[0]);
        const [bNum, bOk] = toFloat(a[1]);
        if (!aOk || !bOk || bNum === 0) {
            return null;
        }
        return aNum / bNum;
    },

    // === Date Operators ===
    'before': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        const [aDate, aOk] = parseDate(a[0]);
        const [bDate, bOk] = parseDate(a[1]);
        if (!aOk || !bOk) {
            return false;
        }
        return aDate.getTime() < bDate.getTime();
    },

    'after': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        const [aDate, aOk] = parseDate(a[0]);
        const [bDate, bOk] = parseDate(a[1]);
        if (!aOk || !bOk) {
            return false;
        }
        return aDate.getTime() > bDate.getTime();
    },

    // === Collection Operators ===
    'in': (args, state, resolve) => {
        const a = resolveArgs(args, 2, resolve, state);
        const needle = a[0];
        const haystack = a[1];

        if (needle === null || needle === undefined || haystack === null || haystack === undefined) {
            return false;
        }

        if (Array.isArray(haystack)) {
            for (const item of haystack) {
                if (compareEqual(needle, item)) {
                    return true;
                }
            }
            return false;
        }

        if (typeof haystack === 'string' && typeof needle === 'string') {
            return haystack.includes(needle);
        }

        return false;
    },

    'some': (args, state, resolve) => {
        if (!Array.isArray(args) || args.length < 2) {
            return false;
        }

        // First arg is the array (resolve it)
        const collection = resolve(args[0], state);
        if (!Array.isArray(collection) || collection.length === 0) {
            return false;
        }

        // Second arg is the condition
        const condition = args[1];

        // Check if any element satisfies the condition
        for (const item of collection) {
            const oldContext = state.currentElement;
            state.currentElement = item;
            const result = isTruthy(resolve(condition, state));
            state.currentElement = oldContext;
            if (result) {
                return true;
            }
        }
        return false;
    },

    'all': (args, state, resolve) => {
        if (!Array.isArray(args) || args.length < 2) {
            return false;
        }

        // First arg is the array (resolve it)
        const collection = resolve(args[0], state);
        if (!Array.isArray(collection)) {
            return false;
        }

        // Empty array returns true for "all" (vacuous truth)
        if (collection.length === 0) {
            return true;
        }

        // Second arg is the condition
        const condition = args[1];

        // Check if all elements satisfy the condition
        for (const item of collection) {
            const oldContext = state.currentElement;
            state.currentElement = item;
            const result = isTruthy(resolve(condition, state));
            state.currentElement = oldContext;
            if (!result) {
                return false;
            }
        }
        return true;
    },

    'none': (args, state, resolve) => {
        if (!Array.isArray(args) || args.length < 2) {
            return false;
        }

        // First arg is the array (resolve it)
        const collection = resolve(args[0], state);
        if (!Array.isArray(collection)) {
            return false;
        }

        // Empty array returns true for "none"
        if (collection.length === 0) {
            return true;
        }

        // Second arg is the condition
        const condition = args[1];

        // Check that no element satisfies the condition
        for (const item of collection) {
            const oldContext = state.currentElement;
            state.currentElement = item;
            const result = isTruthy(resolve(condition, state));
            state.currentElement = oldContext;
            if (result) {
                return false;
            }
        }
        return true;
    },
};

/**
 * Apply an operator to its arguments.
 */
export function applyOperator(
    op: string,
    args: unknown,
    state: EvalState,
    resolve: ResolveFn
): unknown {
    const fn = operators[op];
    if (!fn) {
        // Unknown operator - return null
        return null;
    }
    return fn(args, state, resolve);
}

File: pkg/tenet/operators.go
package tenet

import (
	"fmt"
	"strings"
	"time"
)

// executeOperator handles all JSON-logic operators.
// Returns nil for operations on nil values (nil-safe behavior).
func (e *Engine) executeOperator(op string, args any) any {
	switch op {
	// === Variable Access ===
	case "var":
		path, ok := args.(string)
		if !ok {
			return nil
		}
		return e.getVar(path)

	// === Comparison Operators ===
	case "==":
		a := e.resolveArgs(args, 2)
		return e.compareEqual(a[0], a[1])

	case "!=":
		a := e.resolveArgs(args, 2)
		return !e.compareEqual(a[0], a[1])

	case ">":
		a := e.resolveArgs(args, 2)
		return e.compareNumeric(a[0], a[1], func(x, y float64) bool { return x > y })

	case "<":
		a := e.resolveArgs(args, 2)
		return e.compareNumeric(a[0], a[1], func(x, y float64) bool { return x < y })

	case ">=":
		a := e.resolveArgs(args, 2)
		return e.compareNumeric(a[0], a[1], func(x, y float64) bool { return x >= y })

	case "<=":
		a := e.resolveArgs(args, 2)
		return e.compareNumeric(a[0], a[1], func(x, y float64) bool { return x <= y })

	// === Logical Operators ===
	case "and":
		return e.opAnd(args)

	case "or":
		return e.opOr(args)

	case "not", "!":
		a := e.resolveArgs(args, 1)
		return !e.isTruthy(a[0])

	case "if":
		return e.opIf(args)

	// === Arithmetic Operators ===
	case "+":
		a := e.resolveArgs(args, 2)
		return e.opAdd(a[0], a[1])

	case "-":
		a := e.resolveArgs(args, 2)
		return e.opSubtract(a[0], a[1])

	case "*":
		a := e.resolveArgs(args, 2)
		return e.opMultiply(a[0], a[1])

	case "/":
		a := e.resolveArgs(args, 2)
		return e.opDivide(a[0], a[1])

	// === Date Operators ===
	case "before":
		a := e.resolveArgs(args, 2)
		return e.compareDates(a[0], a[1], func(x, y time.Time) bool { return x.Before(y) })

	case "after":
		a := e.resolveArgs(args, 2)
		return e.compareDates(a[0], a[1], func(x, y time.Time) bool { return x.After(y) })

	// === Collection Operators ===
	case "in":
		a := e.resolveArgs(args, 2)
		return e.opIn(a[0], a[1])

	case "some":
		return e.opSome(args)

	case "all":
		return e.opAll(args)

	case "none":
		return e.opNone(args)

	default:
		// Unknown operator - add error and return nil
		e.addError("", "", ErrRuntimeWarning, fmt.Sprintf("Unknown operator '%s' in logic expression", op), "")
		return nil
	}
}

// === Comparison Helpers ===

// compareEqual checks equality, handling type coercion.
// nil == nil is true, nil == anything_else is false.
func (e *Engine) compareEqual(a, b any) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}

	// Try numeric comparison if both can be numbers
	aNum, aOk := toFloat(a)
	bNum, bOk := toFloat(b)
	if aOk && bOk {
		return aNum == bNum
	}

	// String comparison
	return fmt.Sprintf("%v", a) == fmt.Sprintf("%v", b)
}

// compareNumeric compares two values numerically.
// Returns false if either value is nil or non-numeric.
func (e *Engine) compareNumeric(a, b any, cmp func(float64, float64) bool) bool {
	if a == nil || b == nil {
		return false
	}

	aNum, aOk := toFloat(a)
	bNum, bOk := toFloat(b)
	if !aOk || !bOk {
		return false
	}

	return cmp(aNum, bNum)
}

// compareDates compares two date values.
// Accepts strings (ISO format) or time.Time values.
// Returns false if either value is nil or unparseable.
func (e *Engine) compareDates(a, b any, cmp func(time.Time, time.Time) bool) bool {
	aTime, aOk := parseDate(a)
	bTime, bOk := parseDate(b)
	if !aOk || !bOk {
		return false
	}
	return cmp(aTime, bTime)
}

// === Logical Operators ===

// opAnd returns true if all arguments are truthy.
// Short-circuits on first falsy value.
func (e *Engine) opAnd(args any) bool {
	arr, ok := args.([]any)
	if !ok {
		return e.isTruthy(e.resolve(args))
	}

	for _, arg := range arr {
		if !e.isTruthy(e.resolve(arg)) {
			return false
		}
	}
	return true
}

// opOr returns true if any argument is truthy.
// Short-circuits on first truthy value.
func (e *Engine) opOr(args any) bool {
	arr, ok := args.([]any)
	if !ok {
		return e.isTruthy(e.resolve(args))
	}

	for _, arg := range arr {
		if e.isTruthy(e.resolve(arg)) {
			return true
		}
	}
	return false
}

// opIf implements ternary logic: {"if": [condition, then, else]}
// Also supports chained conditions: {"if": [c1, t1, c2, t2, ..., else]}
func (e *Engine) opIf(args any) any {
	arr, ok := args.([]any)
	if !ok || len(arr) < 2 {
		return nil
	}

	// Process condition-then pairs
	for i := 0; i+1 < len(arr); i += 2 {
		condition := e.resolve(arr[i])
		if e.isTruthy(condition) {
			return e.resolve(arr[i+1])
		}
	}

	// Else clause (odd number of elements = has else)
	if len(arr)%2 == 1 {
		return e.resolve(arr[len(arr)-1])
	}

	return nil
}

// === Arithmetic Operators ===

// opAdd adds two numbers. Returns nil if either is nil.
func (e *Engine) opAdd(a, b any) any {
	if a == nil || b == nil {
		return nil
	}
	aNum, aOk := toFloat(a)
	bNum, bOk := toFloat(b)
	if !aOk || !bOk {
		return nil
	}
	return aNum + bNum
}

// opSubtract subtracts b from a. Returns nil if either is nil.
func (e *Engine) opSubtract(a, b any) any {
	if a == nil || b == nil {
		return nil
	}
	aNum, aOk := toFloat(a)
	bNum, bOk := toFloat(b)
	if !aOk || !bOk {
		return nil
	}
	return aNum - bNum
}

// opMultiply multiplies two numbers. Returns nil if either is nil.
func (e *Engine) opMultiply(a, b any) any {
	if a == nil || b == nil {
		return nil
	}
	aNum, aOk := toFloat(a)
	bNum, bOk := toFloat(b)
	if !aOk || !bOk {
		return nil
	}
	return aNum * bNum
}

// opDivide divides a by b. Returns nil if either is nil or b is zero.
func (e *Engine) opDivide(a, b any) any {
	if a == nil || b == nil {
		return nil
	}
	aNum, aOk := toFloat(a)
	bNum, bOk := toFloat(b)
	if !aOk || !bOk || bNum == 0 {
		return nil
	}
	return aNum / bNum
}

// === Collection Operators ===

// opSome returns true if ANY element in the array satisfies the condition.
// Syntax: {"some": [{"var": "items"}, {"==": [{"var": ""}, "special"]}]}
func (e *Engine) opSome(args any) bool {
	arr, ok := args.([]any)
	if !ok || len(arr) < 2 {
		return false
	}

	// First arg is the array (resolve it)
	collection := e.resolve(arr[0])
	items, ok := collection.([]any)
	if !ok || len(items) == 0 {
		return false
	}

	// Second arg is the condition
	condition := arr[1]

	// Check if any element satisfies the condition
	for _, item := range items {
		if e.evalWithContext(condition, item) {
			return true
		}
	}
	return false
}

// opAll returns true if ALL elements in the array satisfy the condition.
// Syntax: {"all": [{"var": "items"}, {">=": [{"var": ""}, 50]}]}
func (e *Engine) opAll(args any) bool {
	arr, ok := args.([]any)
	if !ok || len(arr) < 2 {
		return false
	}

	// First arg is the array (resolve it)
	collection := e.resolve(arr[0])
	items, ok := collection.([]any)
	if !ok {
		return false
	}

	// Empty array returns true for "all" (vacuous truth)
	if len(items) == 0 {
		return true
	}

	// Second arg is the condition
	condition := arr[1]

	// Check if all elements satisfy the condition
	for _, item := range items {
		if !e.evalWithContext(condition, item) {
			return false
		}
	}
	return true
}

// opNone returns true if NO elements in the array satisfy the condition.
// Syntax: {"none": [{"var": "items"}, {"==": [{"var": ""}, "forbidden"]}]}
func (e *Engine) opNone(args any) bool {
	arr, ok := args.([]any)
	if !ok || len(arr) < 2 {
		return false
	}

	// First arg is the array (resolve it)
	collection := e.resolve(arr[0])
	items, ok := collection.([]any)
	if !ok {
		return false
	}

	// Empty array returns true for "none"
	if len(items) == 0 {
		return true
	}

	// Second arg is the condition
	condition := arr[1]

	// Check that no element satisfies the condition
	for _, item := range items {
		if e.evalWithContext(condition, item) {
			return false
		}
	}
	return true
}

// evalWithContext evaluates a condition with a temporary context value.
// Used by some/all/none to set the current element as {"var": ""}.
func (e *Engine) evalWithContext(condition any, contextValue any) bool {
	// Save and restore the context value for {"var": ""}
	oldContext := e.currentElement
	e.currentElement = contextValue
	result := e.isTruthy(e.resolve(condition))
	e.currentElement = oldContext
	return result
}

// opIn checks if needle is in haystack (array or string).
func (e *Engine) opIn(needle, haystack any) bool {
	if needle == nil || haystack == nil {
		return false
	}

	switch h := haystack.(type) {
	case []any:
		for _, item := range h {
			if e.compareEqual(needle, item) {
				return true
			}
		}
		return false

	case string:
		needleStr, ok := needle.(string)
		if !ok {
			return false
		}
		return strings.Contains(h, needleStr)

	default:
		return false
	}
}

// === Helper Functions ===

// isSlice returns true if the value is a slice/array (e.g. []any from JSON).
func isSlice(v any) bool {
	if v == nil {
		return false
	}
	switch v.(type) {
	case []any, []string, []float64, []int:
		return true
	default:
		return false
	}
}

// toFloat converts a value to float64 if possible.
func toFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case string:
		// Don't auto-convert strings to numbers
		return 0, false
	default:
		return 0, false
	}
}

// parseDate parses a date value (string or time.Time).
// Supports ISO 8601 formats.
func parseDate(v any) (time.Time, bool) {
	if v == nil {
		return time.Time{}, false
	}

	switch d := v.(type) {
	case time.Time:
		return d, true
	case string:
		// Try common formats
		formats := []string{
			time.RFC3339,
			"2006-01-02T15:04:05",
			"2006-01-02",
		}
		for _, format := range formats {
			if t, err := time.Parse(format, d); err == nil {
				return t, true
			}
		}
		return time.Time{}, false
	default:
		return time.Time{}, false
	}
}

